install/install.go

```diff
@@ -1416,15 +1419,83 @@ func (k *K8sInstaller) deployResourceQuotas(ctx context.Context) error {
 	return nil
 }
 
-func (k *K8sInstaller) restartUnmanagedPods(ctx context.Context) error {
-	var printed bool
+func tail(pods []*corev1.Pod) []*corev1.Pod {
+	if len(pods) <= 1 {
+		pods = nil
+	} else {
+		pods = pods[1:]
+	}
+	return pods
+}
+
+func groupPodsByNamespace(pods *corev1.PodList) map[string][]*corev1.Pod {
+	nsToPods := make(map[string][]*corev1.Pod)
+	for idx, p := range pods.Items {
+		nsToPods[p.Namespace] = append(nsToPods[p.Namespace], &pods.Items[idx])
+	}
+	return nsToPods
+}
+
+func (k *K8sInstaller) runEvictions(ctx context.Context, cepMap map[string]struct{}, pods *corev1.PodList) {
+	nsToPods := groupPodsByNamespace(pods)
+	var wg sync.WaitGroup
+	for ns, pods := range nsToPods {
+		ns := ns
+		pods := pods
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			k.Log("♻️  Restarting unmanaged pods on namespace=%s", ns)
+			var retry int
+			maxRetry := 5
+			for len(pods) > 0 {
```

Comment:
I think this loop can be written more clearly as a standard `for` ... `range` loop rather than treating `pods` as a list and replacing it with its tail.

Reply:
Hi @twpayne , thanks for your feedback. I understand that it may seem an exotic way to do a `for` ... `range` loop, but I'm doing it this way because of the nature of the eviction process. The slice we are looping has a set of pods that belong to the same namespace and _likely_ have the same Pod Disruption Budget in common. On a given iteration, you take one pod and it may fail because of PDB contraints, in that case it's of no use to go to the next pod because you're just evicting another replica of the same pod family and you'll be hit again by the PDB contraints. That's why I came up with a stack where we loop from the top and only pop that element (tail) as soon as we make progress on the eviction.  I reckon we could make the code clearer to convey the fact that we are using a stack here.

If I use a `for` ... `range` loop the code becomes confusing and possibly unsafe(?) because for each iteration I may need to re-introduce the same element to ranged over it again. I think that breaks the purpose that a range loop is intended for. 

What do you think ?

Reply:
> I reckon we could make the code clearer to convey the fact that we are using a stack here.

I think that would be quite helpful indeed.

Reply:
Hi @errordeveloper  and @twpayne . I refactor the code and simplify the eviction logic. Instead of grouping pods by namespace I just fire concurrent go routines that would try (and retry) to evict a single pod each. IMHO the code is easier to read and understand. Let me know what you think. Cheers !

Reply:
btw, I will fix the conflicts later.